### PR TITLE
UnmanagedCallersOnlyAttribute Support

### PR DESCRIPTION
**Pull request by joncham**
- Should this pull request have release notes?
  - [X] Yes
  - [ ] No
- Do these changes need to be back ported?
  - [ ] Yes
  - [X] No
- Do these changes need to be upstreamed to [mono/mono](https://github.com/mono/mono) or [dotnet/runtime](https://github.com/dotnet/runtime) repositories?
  - [ ] Yes
  - [X] No

Reviewers: please consider these questions as well! :heart:

**Release notes**

Internal @joncham:
Mono: Add UnmanagedCallersOnlyAttribute support

### CODE REVIEW

File: mono/metadata/marshal.c [hunk 1]
```diff
@@ -6518,7 +6603,34 @@ gboolean
 mono_method_has_unmanaged_callers_only_attribute (MonoMethod *method)
 {
 #ifndef ENABLE_NETCORE
-	return FALSE;
+	ERROR_DECL (attr_error);
+	MonoClass* attr_klass = NULL;
+	/* attribute is only allowed on static methods */
+	if (!m_method_is_static(method))
```
Comment (joncham) [inline mono/metadata/marshal.c]: Should we keep these early outs to speed up this check? Or should we let everything through and allow later "error" checks to handle these cases?

**Comment by UnityAlex** [comment]
Judging by the NETCORE checks I am guessing some of this is used upstream already? Just trying to get a feel for if this is used in production already or not.

**Comment by joncham** [comment]
> Judging by the NETCORE checks I am guessing some of this is used upstream already? Just trying to get a feel for if this is used in production already or not.

The first commit I made myself. The second commit was inspired by upstream but quite modified to avoid taking a bunch of changes.

File: mono/metadata/marshal.c [hunk 2]
```diff
@@ -3264,6 +3265,88 @@ mono_marshal_set_callconv_from_modopt (MonoMethod *method, MonoMethodSignature *
...
+#ifdef ENABLE_NETCORE
+	MonoClass* attr_class = mono_class_try_get_unmanaged_callers_only_attribute_class ();
+	if (!attr_class)
+		return;
+#endif
+
+	ERROR_DECL(error);
+	MonoCustomAttrInfo* cinfo = mono_custom_attrs_from_method_checked (method, error);
+	if (!is_ok(error) || !cinfo) {
+		mono_error_cleanup(error);
```
Comment (UnityAlex) [inline mono/metadata/marshal.c:3308]: Doesn't look like `mono_error_cleanup` is called if we get past here.

File: mono/metadata/marshal.c (hunk 1 above)
Comment (UnityAlex) [reply mono/metadata/marshal.c]: I'm fine with keeping the early outs.

File: mono/metadata/marshal.c [hunk 3]
```diff
@@ -3264,6 +3265,88 @@ mono_marshal_set_callconv_from_modopt (MonoMethod *method, MonoMethodSignature *
...
+	else if (!strcmp (name, "Stdcall"))
+		sig->call_convention = MONO_CALL_STDCALL;
+	else if (!strcmp (name, "Thiscall"))
+		sig->call_convention = MONO_CALL_THISCALL;
+	else if (!strcmp (name, "Fastcall"))
+		sig->call_convention = MONO_CALL_FASTCALL;
+	else if (!strcmp (name, "SuppressGCTransition"))
+		sig->suppress_gc_transition = 1;
+	// TODO: Support CallConvMemberFunction?
+	// TODO: Support multiple calling convetions?
```
Comment (scott-ferguson-unity) [inline mono/metadata/marshal.c:3291]: I played multiple calling conventions in CoreCLR and it only supports multiple calling conventions when they don't conflict.  e.g. CallConvStdCall + CallConvMemberFunction is OK, but CallConvStdCall + CallConvCdecl throws:
> System.InvalidProgramException: 'Multiple unmanaged calling conventions are specified. Only a single calling convention is supported.'

File: mono/metadata/marshal.c (hunk 1 above)
Comment (scott-ferguson-unity) [reply mono/metadata/marshal.c]: Roslyn will error on these cases too, so only users doing bad custom IL should hit these failure cases, so as long as we do something other than randomly crash in hard to debug ways we should be fine.  What is the failure mode in those cases?

**Comment by scott-ferguson-unity** [comment]
Also `UnmanagedCallersOnlyAttribute` isn't defined in our classlibs - we'll have to add it too.

**Comment by joncham** [comment]
> Also `UnmanagedCallersOnlyAttribute` isn't defined in our classlibs - we'll have to add it too.

Why do we need to add it?

**Review by mrvoorhe — CHANGES_REQUESTED** [review]
Changes are needed to make this work with all managed code stripping levels.

coreclr ensures that `UnmanagedCallersOnlyAttribute` survives using the embedded linker descriptor file in `System.Private.CoreLib` https://github.com/Unity-Technologies/runtime/blob/1cd7eb4816b773f35cc75e522e4339525a2a8c51/src/mono/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.xml#L635

We don't have that logic.

And if we don't add the attribute to our class libraries, then we wouldn't be able to mirror the approach used by coreclr.  Instead we'd have to add logic to UnityLinker.

**Review by mrvoorhe — CHANGES_REQUESTED** [review]
Changes are needed to make this work with all managed code stripping levels.

coreclr ensures that `UnmanagedCallersOnlyAttribute` survives using the embedded linker descriptor file in `System.Private.CoreLib` https://github.com/Unity-Technologies/runtime/blob/1cd7eb4816b773f35cc75e522e4339525a2a8c51/src/mono/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.xml#L635

We don't have that in our [mscorlib.xml](https://github.com/Unity-Technologies/mono/blob/unity-main/mcs/class/corlib/LinkerDescriptor/mscorlib.xml)

And if we don't add the attribute to our class libraries, then we wouldn't be able to mirror the approach used by coreclr.  Instead we'd have to add logic to UnityLinker.

**Comment by joncham** [comment]
@mrvoorhe I'm not sure what the requested change is? We need to support this attribute defined by anyone in any assembly, similar to `MonoPInvokeCallbackAttribute`.

**Comment by mrvoorhe** [comment]
> @mrvoorhe I'm not sure what the requested change is? We need to support this attribute defined by anyone in any assembly, similar to `MonoPInvokeCallbackAttribute`.

If you want to go that route that is fine.  In that case no changes would be needed to this PR, but a PR to UnityLinker should be made to add the same logic as `MonoPInvokeCallbackAttribute`


https://github.cds.internal.unity3d.com/unity/il2cpp/blob/main/UnityLinker/Linker.Steps/Marking/CommonRuntimeMarking.cs
https://github.cds.internal.unity3d.com/unity/il2cpp/blob/main/Unity.Linker.Tests.Cases/Il2CppRuntime/Attributes/OnlyKeepUsed/MonoPInvokeCallbackAttributeIsKept.cs
https://github.cds.internal.unity3d.com/unity/il2cpp/blob/main/Unity.Linker.Tests.Cases/MonoRuntime/Attributes/OnlyKeepUsed/MonoPInvokeCallbackAttributeIsKept.cs

**Review by mrvoorhe — APPROVED** [review]

**Comment by scott-ferguson-unity** [comment]
> > Also `UnmanagedCallersOnlyAttribute` isn't defined in our classlibs - we'll have to add it too.
> 
> Why do we need to add it?

Wouldn't we need to to compile/resolve any of the Mono/IL2CPP usages of the attribute we're planning to add?  Or can we leave it out as long as we're doing IL injection?

File: mono/metadata/marshal.c (hunk 1 above)
Comment (bholmes) [reply mono/metadata/marshal.c]: Should the early outs be in mono_marshal_set_callconv_from_unmanaged_callers_only_attribute as well. 

I feel that mono_method_has_unmanaged_callers_only_attribute and mono_marshal_set_callconv_from_unmanaged_callers_only_attribute duplicate code and are not consistent?

**Review by bholmes — APPROVED** [review]
If the testing looks good, ship it!

**Comment by joncham** [comment]
> Wouldn't we need to to compile/resolve any of the Mono/IL2CPP usages of the attribute we're planning to add? Or can we leave it out as long as we're doing IL injection?

We can add it if the class libraries themselves need it. But for Unity usage, we already define it in UnityEngine for when we target CoreCLR. We will match on name and not assembly (like we do other attributes).

Once we target .NET as a compilation target (instead of .NET Standard and .NET Framework+) we can use the one provided by the profile itself.

File: mono/metadata/marshal.c (hunk 1 above)
Comment (joncham) [reply mono/metadata/marshal.c]: Ideally, I would not have early outs. I'm just a bit worried about the cost of the attribute checks. Let me evaluate any impact again, and try to just leave a regular code path.

Comment (joncham) [reply mono/metadata/marshal.c]: I have removed the early outs, this means we will hit error code paths at other places in mono code. For example, when the ldftn is encountered it validates proper use of the attribute.

If we have specific error cases we care about, we should write tests in IL2CPP or Unity, and ensure all runtimes properly handle them.

File: mono/metadata/marshal.c (hunk 2 above)
Comment (joncham) [reply mono/metadata/marshal.c:3308]: I think that only needs called when an error happens. I have added checks after each function taking an error, and just `mono_error_assert_ok` for now. I'd rather crash and detect we are doing something wrong I think.

**Review by UnityAlex — APPROVED** [review]

**Pull request merged**